packages/store-engine/src/main/engine/FileEngine.ts

```diff
@@ -36,32 +38,36 @@ export default class FileEngine implements CRUDEngine {
     return fs.remove(this.storeName);
   }
 
-  private resolvePath(tableName: string, primaryKey?: string): Promise<string> {
-    const isPathTraversal = (...testPaths: string[]): boolean => {
-      for (const testPath of testPaths) {
-        if (
-          typeof testPath !== 'undefined' &&
-          (testPath.includes('.') || testPath.includes('/') || testPath.includes('\\'))
-        ) {
-          return true;
-        }
-      }
-      return false;
-    };
+  static enforcePathRestrictions(givenTrustedRoot: string, givenPath: string, forceWindows: boolean = false): void {
+    const pathApi: any = process.platform === 'win32' || forceWindows === true ? path.win32 : path;
+    const trustedRoot = pathApi.resolve(givenTrustedRoot);
+
+    const trustedRootDetails = pathApi.parse(trustedRoot);
+    if (trustedRootDetails.root === trustedRootDetails.dir && trustedRootDetails.base === '') {
+      const message = `"${trustedRoot}" cannot be the root of the filesystem.`;
+      throw new PathValidationError(message);
+    }
 
+    const unsafePath = pathApi.resolve(pathApi.join(trustedRoot, givenPath));
+    if (unsafePath.startsWith(trustedRoot) === false) {
+      const message = `Path traversal has been detected. Allowed path was "${trustedRoot}" but tested path "${givenPath}" attempted to reach "${unsafePath}"`;
+      throw new PathValidationError(message);
+    }
+  }
+
+  private resolvePath(tableName: string, primaryKey: string = ''): Promise<string> {
     return new Promise((resolve, reject) => {
-      if (isPathTraversal(tableName, primaryKey || '')) {
-        const message = `Path traversal has been detected on "${path.join(tableName, String(primaryKey))}".`;
-        return reject(new PathValidationError(message));
-      }
+      FileEngine.enforcePathRestrictions(this.baseDirectory, this.storeNameOriginal);
```

Comment:
Executing `FileEngine.enforcePathRestrictions` three times on every `resolvePath` sounds too much to me. Keep in mind that `resolvePath` is executed on every CRUD operation. So every "read" will cause such excessive checks which will make our code slow. I think we can come up with something more performant.